pkg/microservice/policy/core/repository/mongodb/policy.go

```diff
@@ -0,0 +1,195 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mongodb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/koderover/zadig/pkg/config"
+	"github.com/koderover/zadig/pkg/microservice/policy/core/repository/models"
+	mongotool "github.com/koderover/zadig/pkg/tool/mongo"
+)
+
+type PolicyColl struct {
+	*mongo.Collection
+
+	coll string
+}
+
+func NewPolicyColl() *PolicyColl {
+	name := models.Policy{}.TableName()
+	return &PolicyColl{
+		Collection: mongotool.Database(config.MongoDatabase()).Collection(name),
+		coll:       name,
+	}
+}
+
+func (c *PolicyColl) GetCollectionName() string {
+	return c.coll
+}
+
+func (c *PolicyColl) EnsureIndex(ctx context.Context) error {
+	mod := mongo.IndexModel{
+		Keys: bson.D{
+			bson.E{Key: "namespace", Value: 1},
+			bson.E{Key: "name", Value: 1},
+		},
+		Options: options.Index().SetUnique(true),
+	}
+
+	_, err := c.Indexes().CreateOne(ctx, mod)
+
+	return err
+}
+
+func (c *PolicyColl) Get(ns, name string) (*models.Policy, bool, error) {
+	res := &models.Policy{}
+
+	query := bson.M{"namespace": ns, "name": name}
+	err := c.FindOne(context.TODO(), query).Decode(res)
+	if err != nil {
+		if err == mongo.ErrNoDocuments {
+			return nil, false, nil
+		}
+
+		return nil, false, err
+	}
+
+	return res, true, nil
+}
+
+func (c *PolicyColl) List() ([]*models.Policy, error) {
+	var res []*models.Policy
+
+	ctx := context.Background()
+
+	cursor, err := c.Collection.Find(ctx, bson.M{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = cursor.All(ctx, &res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (c *PolicyColl) ListBy(projectName string) ([]*models.Policy, error) {
+	var res []*models.Policy
+
+	ctx := context.Background()
+	query := bson.M{"namespace": projectName}
+
+	cursor, err := c.Collection.Find(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	err = cursor.All(ctx, &res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (c *PolicyColl) ListBySpaceAndName(projectName string, name string) ([]*models.Policy, error) {
+	var res []*models.Policy
+
+	ctx := context.Background()
+	query := bson.M{"namespace": projectName, "name": name}
+
+	cursor, err := c.Collection.Find(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	err = cursor.All(ctx, &res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
```

Comment:
why not use a  parameter holds the query option to reduce ```List() ListBy() ListBySpaceAndName() ``` functions to one, just like other models?

Reply:
when opt to much , it will only increase complexity

Reply:
i don't see any complexity of adding a query option only contains two parameters: `namespace ` and `name`, writing one function for each query condition is another kind of `complexity`